modules/calib3d/src/hal_replacement.hpp

```diff
@@ -0,0 +1,132 @@
+// are permitted provided that the following conditions are met:
+//
+//   * Redistribution's of source code must retain the above copyright notice,
+//     this list of conditions and the following disclaimer.
+//
+//   * Redistribution's in binary form must reproduce the above copyright notice,
+//     this list of conditions and the following disclaimer in the documentation
+//     and/or other materials provided with the distribution.
+//
+//   * The name of the copyright holders may not be used to endorse or promote products
+//     derived from this software without specific prior written permission.
+//
+// This software is provided by the copyright holders and contributors "as is" and
+// any express or implied warranties, including, but not limited to, the implied
+// warranties of merchantability and fitness for a particular purpose are disclaimed.
+// In no event shall the Intel Corporation or contributors be liable for any direct,
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#ifndef OPENCV_CALIB3D_HAL_REPLACEMENT_HPP
+#define OPENCV_CALIB3D_HAL_REPLACEMENT_HPP
+
+#include "opencv2/core/hal/interface.h"
+
+#if defined(__clang__)  // clang or MSVC clang
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-parameter"
+#elif defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable : 4100)
+#elif defined(__GNUC__)
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-parameter"
+#endif
+
+//! @addtogroup calib3d_hal_interface
+//! @note Define your functions to override default implementations:
+//! @code
+//! #undef hal_add8u
+//! #define hal_add8u my_add8u
+//! @endcode
+//! @{
+
+/**
+   @brief Project points from 3D world space to 2D screen space using rotation and translation matrix and camera intrinsic parameters
+   @param src_data Pointer to source 3D points array
+   @param src_step Step between consecutive points in 3D points array
+   @param src_size Amount of points in 3D points array
+   @param dst_data Pointer to resulting 2D points array
+   @param dst_step Step between consecutive points in 2D resulting array
+   @param rt_data Pointer to 3x4 array containing rotation-then-translation matrix
+   @param intrinsics_data Pointer to camera intrinsic parameters vector containing [fx, fy, cx, cy, k0, k1, k2, k3]
+*/
+inline int hal_ni_projectpoints32f(const float* src_data, uint32_t src_step, uint32_t src_size, float* dst_data, uint32_t dst_step, const float* rt_data, const float* intrinsics_data) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
```

Comment:
sometimes it's convenient to store input and/or output points as structure of arrays, not array of structures. I think, HAL function should support both. Can we have:

```
inline int hal_ni_projectpoints32f(const float* X_data, const float* Y_data, const float* Z_data, size_t XYZ_step, size_t src_size, float* x_data, float* y_data, size_t dst_step, const float* rt_data, const float* intrinsics_data, int flags) { return CV_HAL_ERROR_NOT_IMPLEMENTED; }
```

to let user pass separate pointers? Interleaved data is also supported

Reply:
1. About separate pointers for separate coordinates: this is a topic for discussion.
 * To my mind, we need to support two most popular layouts: `xyzxyzxyz...` and `xxx...yyy...zzz...`, other ones seem too exotic to me.
 * We can do this by providing two strides: step1 between X, Y and Z coordinates of the same point and step2 between consecutive points.
2. I think we need to pass all the distortion coefficients from OpenCV to HAL in the same order. We don't even have to provide the number of actual coefficients passed because the coefficients have no effect if they are zero.
3. If we stick to pinhole camera model (which is what everybody expects from `projectPoints`) I think we may consider it future-proof-enough since the camera model itself does not change. Other camera models (e.g. fisheye) should have different HAL calls, there's no sense to join every `projectPoints` function under the only HAL call.